rust/src/linearview.rs

```diff
@@ -0,0 +1,515 @@
+        }
+    }
+}
+
+unsafe impl RefCountable for LinearViewCursor {
+    unsafe fn inc_ref(handle: &Self) -> Ref<Self> {
+        Ref::new(Self {
+            handle: BNNewLinearViewCursorReference(handle.handle),
+        })
+    }
+
+    unsafe fn dec_ref(handle: &Self) {
+        BNFreeLinearViewCursor(handle.handle);
+    }
+}
+
+impl ToOwned for LinearViewCursor {
+    type Owned = Ref<Self>;
+
+    fn to_owned(&self) -> Self::Owned {
+        unsafe { RefCountable::inc_ref(self) }
+    }
+}
+
+unsafe impl Send for LinearViewCursor {}
+unsafe impl Sync for LinearViewCursor {}
+
+pub struct LinearViewLinesIterator {
+    count: usize,
+    lines: *mut BNLinearDisassemblyLine,
+    offset: usize,
+}
+
+impl LinearViewLinesIterator {
+    pub(crate) fn new(count: usize, lines: *mut BNLinearDisassemblyLine) -> Self {
+        Self {
+            count,
+            lines,
+            offset: 0,
+        }
+    }
+
+    pub(crate) fn empty() -> Self {
+        Self {
+            count: 0,
+            lines: std::ptr::null_mut(),
+            offset: 0,
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.count
+    }
+}
+
+impl Iterator for LinearViewLinesIterator {
+    type Item = LinearDisassemblyLine;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset >= self.count {
+            return None;
+        }
+        let line = unsafe { LinearDisassemblyLine::from_raw(self.lines.offset(self.offset as _)) };
+
+        self.offset += 1;
+        Some(line)
+    }
+}
+
+
+// FIXME: we get a segmentation fault if this is there. However,
+// I don't understand enough of the API to see where this should get freed otherwise.
+/*
+impl Drop for LinearViewLinesIterator {
+    fn drop(&mut self) {
+         unsafe {
+             BNFreeLinearDisassemblyLines(self.lines, self.count);
+        }
+    }
+}
+*/
+
+pub type LinearDisassemblyLineType = BNLinearDisassemblyLineType;
+
+pub struct LinearDisassemblyLine {
+    r#type: LinearDisassemblyLineType,
+    function: Ref<Function>,
+    contents: BNDisassemblyTextLine,
+}
+
+impl LinearDisassemblyLine {
+    pub(crate) unsafe fn from_raw(raw: *mut BNLinearDisassemblyLine) -> Self {
+        let line = *raw;
+        let linetype = line.type_;
+        let function = Function::from_raw(line.function);
+        let contents = line.contents;
+        Self {
+            r#type: linetype,
+            function,
+            contents,
+        }
+    }
+
+    pub fn addr(&self) -> u64 {
+        self.contents.addr
+    }
+
+    pub fn instr_idx(&self) -> usize {
+        self.contents.instrIndex
+    }
+
+    pub fn count(&self) -> usize {
+        self.contents.count
+    }
+
+    pub fn tag_count(&self) -> usize {
+        self.contents.tagCount
+    }
+
+    pub fn tokens(&self) -> LinearDisassemblyLineTokenIterator {
+        LinearDisassemblyLineTokenIterator::new(self.contents.count, self.contents.tokens)
+    }
+
+    pub fn function(&self) -> Ref<Function> {
```

Comment:
It might be better to return a `&Function` which they can `to_owned` if they need to.  I'll accept this as is though, since there are other parts of the API where we do exactly this.